keps/sig-cli/20190805-kubectl-debug.md

```diff
@@ -0,0 +1,460 @@
+  -i, --stdin=true: Pass stdin to the container
+  --image='': Required. Container image to use for debug container.
+  --target='': Target processes in this container name.
+  -t, --tty=true: Stdin is a TTY
+
+Usage:
+  kubectl debug (POD | TYPE/NAME) [-c CONTAINER] [flags] -- COMMAND [args...] [options]
+
+Use "kubectl options" for a list of global command-line options (applies to all commands).
+```
+
+### Pod Troubleshooting by Copy
+
+Pod troubleshooting via Ephemeral Container relies on an alpha feature which is
+unlikely to be enabled on production clusters. In order to support these use
+cases, and because it would be generally useful, we will support a mode of Pod
+troubleshooting that behaves similar to Pod Troubleshooting with Ephemeral
+Debug Container but operates instead on a copy of the target pod.
+
+The following additional options will cause a copy of the target pod to be
+created:
+
+```
+Options:
+  --copy-to='': Create a copy of the target pod with this pod name.
+  --delete-old: When used with `--copy-to`, delete the original pod.
+  --copy-labels=false: When used with `--copy-to`, specifies whether labels
+                       should also be copied. Note that copying labels may cause
+                       the copy to receive traffic from a service or a replicaset
+                       to kill other pods.
+  --share-processes=true: When used with `--copy-to`, enable process namespace
+                          sharing in the pod copy.
+```
+
+The modification `kubectl debug` makes to `Pod.Spec.Containers` depends on the
+value of the `--container` flag.
+
+#### Creating a Debug Container by copy
+
+If a user does not specify a `--container` or specifies one that does not exist,
+then the user is instructing `kubectl debug` to create a new Debug Container in
+the Pod copy.
+
+```
+Examples:
+  # Create a copy of 'mypod' with a new debugging container and attach to it
+  kubectl debug mypod --copy-to=mypod-debug --image=debian --attach
+```
+
+#### Modify Application Image by Copy
+
+If a users specify a `--container`, then they are instructing `kubectl debug` to
+create a copy of the target pod with a new image for one of the containers.
+
+```
+Examples:
+  # Create a copy of 'mypod' with the debugging image for container 'app'
+  kubectl debug mypod --copy-to=mypod-debug --image=myapp-image:debug --container=myapp
+```
+
+Note that the Pod API allows updates of container images in-place, so
+`--copy-to` is not necessary for this operation. `kubectl debug` isn't necessary
+to achieve this -- it can be done today with patch -- but `kubectl debug` could
+implement it as well for completeness.
+
+### Node Troubleshooting with Privileged Containers
+
+When invoked with a node as a target, `kubectl debug` will create a new
+privileged pod constrained to this particular node and run in the host
+namespaces.
+
+TODO: how?
+
+TODO: Some CLI options only make sense for pods, and some only make sense for 
```

Comment:
I'm not strongly convinced, but apiserver did something https://github.com/kubernetes/kubernetes/pull/64517 that I was planning o implementing in kubectl, eventually. But let's start simple.